File: sdk-go/dex/attribute.go
// Legacy Materials in this file remain under their original licenses.
// See LEGACY_NOTICES.md.

// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications after the Legacy Cutoff are licensed under the
// Super Durable Source License 1.0.
// Legacy Materials remain under their original licenses.
// See LICENSE and LEGACY_NOTICES.md.

package dex

import "github.com/superdurable/dex/sdk-go/gen/dexpb"

// Attribute defines a typed persisted value.
//
// Declare an Attribute once, include it in the Flow's PersistenceSchema, and reuse the same value
// from Step and RPC code. Values use the SDK's primitive and JSON encoding rules.
// String values require valid UTF-8; use []byte for arbitrary bytes.
//
// Example:
//
//	var orderStatus = dex.DefineAttribute[string](
//		"order-status",
//		dex.Indexed(dex.AttributeIndex{Type: dex.IndexKeyword}),
//	)
type Attribute[T any] struct {
	name                 string
	index                *AttributeIndex
	syncToAttributeStore bool
}

// DefineAttribute creates a typed Attribute with a stable name and definition options.
// "/" is reserved as the AttributeMap separator and is prohibited in Attribute names.
// The definition does not perform I/O; register it through PersistenceSchema before use.
func DefineAttribute[T any](key string, options ...AttributeOption) Attribute[T] {
	config := applyAttributeOptions(options)
	return Attribute[T]{
		name:                 key,
		index:                config.index,
		syncToAttributeStore: config.syncToAttributeStore,
	}
}

// AttributeDef is the interface of Attribute, without Go's generic
// So that internal sdk can use it to workaround Go's generic limitations
//
// AttributeDef is internal to the SDK. Applications create values with DefineAttribute or
// DefineAttributeMap, then pass them to PersistenceSchema and Client methods.
type AttributeDef interface {
	attributeName() string
	attributeIndex() *AttributeIndex
	attributeIsMap() bool
	attributeSyncToAttributeStore() bool
}

// Get returns the current value. Missing values return AttributeNotFoundError.
// It returns an error when ctx is not an SDK invocation or the value cannot be decoded.
func (a Attribute[T]) Get(ctx Context) (value T, err error) {
	invocation, ok := ctx.(attributeInvocation)
	if !ok {
		return value, errInvalidInvocationContext
	}
	found, err := invocation.getAttribute(a.name, &value)
	if err != nil {
		return value, err
	}
	if !found {
		return value, &AttributeNotFoundError{AttributeName: a.name}
	}
	return value, nil
}

// Set stages value for durable persistence when the current Step or RPC invocation succeeds.
// It returns an error when ctx is not an SDK invocation or value cannot be encoded or indexed.
func (a Attribute[T]) Set(ctx Context, value T) error {
	invocation, ok := ctx.(attributeInvocation)
	if !ok {
		return errInvalidInvocationContext
	}
	return invocation.setAttribute(a.name, value, a.index)
}

// Delete stages removal of this Attribute. Deleting a missing value is valid.
// It returns an error when ctx is not an SDK invocation.
func (a Attribute[T]) Delete(ctx Context) error {
	invocation, ok := ctx.(attributeInvocation)
	if !ok {
		return errInvalidInvocationContext
	}
	return invocation.deleteAttribute(a.name, a.index)
}

// AttributeName returns the stable name sent to Dex and used in PersistenceSchema.
func (a Attribute[T]) AttributeName() string {
	return a.name
}

func (a Attribute[T]) attributeName() string {
	return a.name
}

func (a Attribute[T]) attributeIndex() *AttributeIndex {
	return a.index
}

func (Attribute[T]) attributeIsMap() bool {
	return false
}

func (a Attribute[T]) attributeSyncToAttributeStore() bool {
	return a.syncToAttributeStore
}

// AttributeMap defines keyed typed persisted values.
//
// Register the map once in PersistenceSchema, then supply an instance string for every access and lock.
// Slash is prohibited in instance keys because it is a reserved character.
// String values require valid UTF-8; use []byte for arbitrary bytes.
type AttributeMap[T any] struct {
	name                 string
	index                *AttributeIndex
	syncToAttributeStore bool
}

// AttributeMapLoad selects AttributeMap entries for an RPC invocation.
// Create values with AttributeMap.Load or AttributeMap.LoadAll and place them in
// InvokeOptions.LoadAttributeMaps.
type AttributeMapLoad struct {
	name     string
	instance string
	isAll    bool
}

// LoadAll selects every current instance for an RPC snapshot.
// Loading does not lock the map or prevent concurrent changes.
func (a AttributeMap[T]) LoadAll() AttributeMapLoad {
	return AttributeMapLoad{name: a.name, isAll: true}
}

// Load selects one logical instance for an RPC snapshot.
// The SDK escapes instance when constructing the protocol selector. Loading does not lock the
// instance or prevent concurrent changes.
func (a AttributeMap[T]) Load(instance string) AttributeMapLoad {
	return AttributeMapLoad{name: a.name, instance: instance}
}

// DefineAttributeMap creates a typed Attribute map with a stable name and definition options.
// "/" is reserved as the AttributeMap separator and is prohibited in AttributeMap names.
// The definition performs no I/O; register it through PersistenceSchema before use.
func DefineAttributeMap[T any](name string, options ...AttributeOption) AttributeMap[T] {
	config := applyAttributeOptions(options)
	return AttributeMap[T]{
		name:                 name,
		index:                config.index,
		syncToAttributeStore: config.syncToAttributeStore,
	}
}

// Get returns the current map value. Missing instances return AttributeNotFoundError.
// It returns an error when ctx is invalid or the value cannot be decoded.
func (a AttributeMap[T]) Get(
	ctx Context,
	instance string,
) (value T, err error) {
	invocation, ok := ctx.(attributeInvocation)
	if !ok {
		return value, errInvalidInvocationContext
	}
	found, err := invocation.getAttributeMap(a.name, instance, &value)
	if err != nil {
		return value, err
	}
	if !found {
		return value, &AttributeNotFoundError{
			AttributeName: a.name,
			Instance:      instance,
		}
	}
	return value, nil
}

// Set stages value for one map instance when the invocation succeeds.
// It returns an error for an invalid context or a value that cannot be encoded or indexed.
func (a AttributeMap[T]) Set(ctx Context, instance string, value T) error {
	invocation, ok := ctx.(attributeInvocation)
	if !ok {
		return errInvalidInvocationContext
	}
	return invocation.setAttributeMap(a.name, instance, value, a.index)
}

// Delete stages removal of one map instance. Deleting a missing instance is valid.
// It returns an error when ctx is not an SDK invocation.
func (a AttributeMap[T]) Delete(ctx Context, instance string) error {
	invocation, ok := ctx.(attributeInvocation)
	if !ok {
		return errInvalidInvocationContext
	}
	return invocation.deleteAttributeMap(a.name, instance, a.index)
}

// MapSize returns the number of existing instances, including writes buffered by this invocation.
func (a AttributeMap[T]) MapSize(ctx Context) int {
	return len(a.AllInstanceKeys(ctx))
}

// AllInstanceKeys returns existing instance keys in ascending order, including buffered writes.
func (a AttributeMap[T]) AllInstanceKeys(ctx Context) []string {
	invocation, ok := ctx.(attributeInvocation)
	if !ok {
		panic(errInvalidInvocationContext)
	}
	return invocation.attributeMapKeys(a.name)
}

// AttributeName returns the stable shared name of this Attribute map.
func (a AttributeMap[T]) AttributeName() string {
	return a.name
}

func (a AttributeMap[T]) attributeName() string {
	return a.name
}

func (a AttributeMap[T]) attributeIndex() *AttributeIndex {
	return a.index
}

func (AttributeMap[T]) attributeIsMap() bool {
	return true
}

func (a AttributeMap[T]) attributeSyncToAttributeStore() bool {
	return a.syncToAttributeStore
}

// AttributeOption configures an Attribute or AttributeMap definition.
// Use Indexed and SyncToAttributeStore to create values; custom implementations are not supported.
type AttributeOption interface {
	applyAttribute(*attributeConfig)
}

// IndexType identifies the search representation required for an indexed Attribute value.
type IndexType uint8

const (
	// IndexKeyword indexes one exact UTF-8 string for equality and aggregation queries.
	IndexKeyword IndexType = iota + 1
	// IndexFullText indexes analyzed UTF-8 text for full-text queries.
	IndexFullText
	// IndexKeywordArray indexes a slice or array of exact UTF-8 strings.
	IndexKeywordArray
	// IndexInt indexes a signed or unsigned integer that fits in int64.
	IndexInt
	// IndexDouble indexes a finite floating-point value.
	IndexDouble
	// IndexBool indexes a Boolean value.
	IndexBool
	// IndexDatetime indexes time.Time or an RFC3339Nano string.
	IndexDatetime
)

// AttributeIndex configures visibility indexing; datetime values use time.Time or RFC3339Nano strings.
// An empty IndexKey uses the Attribute name
// for single values and a derived physical name for map instances.
type AttributeIndex struct {
	// Type selects the required value representation.
	Type IndexType
	// IndexKey overrides the physical search field; empty uses the Attribute-derived key.
	IndexKey string
}

// Indexed enables search indexing with index on an Attribute or AttributeMap definition.
func Indexed(index AttributeIndex) AttributeOption {
	return indexedAttributeOption{index: index}
}

// SyncToAttributeStore projects every write through the Flow's configured Attribute Store.
// Projection is asynchronous and latest-state only. Deletes write SQL NULL, and failures do not roll back Flow Attributes.
func SyncToAttributeStore() AttributeOption {
	return syncToAttributeStoreOption{}
}

// AttributeLock identifies one Attribute or Attribute-map instance lock.
// Create locks with LockAttribute or LockAttributeMap, then place them in StepOptions or InvokeOptions.
type AttributeLock interface {
	attributeLock()
}

// LockAttribute creates a lock for the single value represented by attribute.
func LockAttribute[T any](attribute Attribute[T]) AttributeLock {
	return attributeLock{name: attribute.name}
}

// LockAttributeMap creates a lock scoped to one Attribute-map instance.
func LockAttributeMap[T any](
	attribute AttributeMap[T],
	instance string,
) AttributeLock {
	return attributeLock{name: attribute.name, instance: instance, isMap: true}
}

// InitialAttribute encodes a value for StartFlowOptions.Attributes.
// It returns ValueMappingError when the value is unsupported or incompatible with its index.
func InitialAttribute[T any](
	attribute Attribute[T],
	value T,
) (InitialAttributeDef, error) {
	encoded, indexConfig, err := encodeAttributeValue(value, attribute.index)
	if err != nil {
		return nil, err
	}
	return initialAttribute{
		name:                 attribute.name,
		value:                value,
		index:                attribute.index,
		syncToAttributeStore: attribute.syncToAttributeStore,
		encoded:              encoded,
		indexConfig:          indexConfig,
	}, nil
}

// InitialAttributeDef is an encoded Attribute initialization accepted by StartFlowOptions.
// Values are created by InitialAttribute and InitialAttributeMapValue.
type InitialAttributeDef interface {
	initialAttribute()
}

// InitialAttributeMapValue encodes one Attribute-map instance for StartFlowOptions.Attributes.
// The instance identifies the map entry. Slash is prohibited in instance keys because it is a reserved character.
// It returns ValueMappingError when the value is unsupported or incompatible with its index.
func InitialAttributeMapValue[T any](
	attribute AttributeMap[T],
	instance string,
	value T,
) (InitialAttributeDef, error) {
	if err := validateMapInstance(instance); err != nil {
		return nil, err
	}
	encoded, indexConfig, err := encodeAttributeValue(value, attribute.index)
	if err != nil {
		return nil, err
	}
	return initialAttribute{
		name:                 attribute.name,
		instance:             instance,
		value:                value,
		index:                attribute.index,
		isMap:                true,
		syncToAttributeStore: attribute.syncToAttributeStore,
		encoded:              encoded,
		indexConfig:          indexConfig,
	}, nil
}

type attributeInvocation interface {
	getAttribute(name string, valuePtr any) (bool, error)
	setAttribute(name string, value any, index *AttributeIndex) error
	deleteAttribute(name string, index *AttributeIndex) error
	getAttributeMap(name string, instance string, valuePtr any) (bool, error)
	setAttributeMap(name string, instance string, value any, index *AttributeIndex) error
	deleteAttributeMap(name string, instance string, index *AttributeIndex) error
	attributeMapKeys(name string) []string
}

type attributeConfig struct {
	index                *AttributeIndex
	syncToAttributeStore bool
}

func applyAttributeOptions(options []AttributeOption) attributeConfig {
	var config attributeConfig
	for _, option := range options {
		option.applyAttribute(&config)
	}
	return config
}

type indexedAttributeOption struct {
	index AttributeIndex
}

func (option indexedAttributeOption) applyAttribute(config *attributeConfig) {
	index := option.index
	config.index = &index
}

type syncToAttributeStoreOption struct{}

func (syncToAttributeStoreOption) applyAttribute(config *attributeConfig) {
	config.syncToAttributeStore = true
}

type attributeLock struct {
	name     string
	instance string
	isMap    bool
}

func (attributeLock) attributeLock() {}

type initialAttribute struct {
	name                 string
	instance             string
	value                any
	index                *AttributeIndex
	isMap                bool
	syncToAttributeStore bool
	encoded              *dexpb.Value
	indexConfig          *dexpb.IndexConfig
}

func (initialAttribute) initialAttribute() {}

File: sdk-rust/crates/dex-sdk/src/attribute.rs
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

use std::marker::PhantomData;

use dex_protocol::dex::IndexConfig;

use crate::{Context, HandlerError, HandlerResult, Value};

/// Defines one durable, typed value stored with a Flow.
///
/// Declare Attributes on the Flow, add them to [`crate::PersistenceSchema`], and reuse the same
/// definition from Step and RPC code. Values use Serde serialization. `get` distinguishes a missing
/// value with `None`; `get_required` converts absence into [`HandlerError`].
///
/// # Examples
///
/// ```
/// use dex_sdk::{Attribute, AttributeIndex};
///
/// let status = Attribute::<String>::new("status")
///     .indexed(AttributeIndex::keyword());
/// ```
pub struct Attribute<T> {
    name: String,
    index: Option<AttributeIndex>,
    marker: PhantomData<fn() -> T>,
}

impl<T> Attribute<T> {
    /// Defines an unindexed Attribute with stable `name`.
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            index: None,
            marker: PhantomData,
        }
    }

    /// Enables Flow-search indexing with the supplied representation.
    pub fn indexed(mut self, index: AttributeIndex) -> Self {
        self.index = Some(index);
        self
    }

    /// Reads the current value from a Step or RPC invocation.
    ///
    /// Returns `Ok(None)` when no value is stored.
    ///
    /// # Errors
    ///
    /// Returns [`HandlerError`] when the payload cannot be decoded.
    pub fn get(&self, context: &Context) -> HandlerResult<Option<T>>
    where
        T: Value,
    {
        context.get_attribute(self)
    }

    /// Reads the current value and treats absence as an error.
    ///
    /// # Errors
    ///
    /// Returns [`HandlerError`] when the Attribute is missing or cannot be decoded.
    pub fn get_required(&self, context: &Context) -> HandlerResult<T>
    where
        T: Value,
    {
        self.get(context)?
            .ok_or_else(|| HandlerError::new(format!("attribute {} is missing", self.name)))
    }

    /// Stages `value` for durable persistence in the current invocation.
    ///
    /// # Errors
    ///
    /// Returns [`HandlerError`] when the value cannot be encoded or indexed.
    pub fn set(&self, context: &mut Context, value: T) -> HandlerResult<()>
    where
        T: Value,
    {
        context.set_attribute(self, value)
    }

    /// Stages deletion of the current value.
    ///
    /// Deleting a missing Attribute is valid.
    pub fn delete(&self, context: &mut Context) -> HandlerResult<()> {
        context.delete_attribute(self)
    }

    /// Creates a lock request for Step options or an RPC definition.
    pub fn lock(&self) -> AttributeLock {
        AttributeLock {
            attribute: self.name.clone(),
            instance: None,
        }
    }

    pub(crate) fn name(&self) -> &str {
        &self.name
    }

    pub(crate) fn index(&self) -> Option<&AttributeIndex> {
        self.index.as_ref()
    }
}

impl<T> Clone for Attribute<T> {
    fn clone(&self) -> Self {
        Self {
            name: self.name.clone(),
            index: self.index.clone(),
            marker: PhantomData,
        }
    }
}

/// Defines keyed durable values sharing one Attribute name.
///
/// Each `instance` identifies an independent stored value and lock. Add the map definition once to
/// [`crate::PersistenceSchema`]; instance names are supplied at access time.
pub struct AttributeMap<T> {
    name: String,
    index: Option<AttributeIndex>,
    marker: PhantomData<fn() -> T>,
}

impl<T> AttributeMap<T> {
    /// Defines an unindexed Attribute map with stable `name`.
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            index: None,
            marker: PhantomData,
        }
    }

    /// Enables search indexing for every instance using the supplied representation.
    pub fn indexed(mut self, index: AttributeIndex) -> Self {
        self.index = Some(index);
        self
    }

    /// Reads one instance, returning `Ok(None)` when it is absent.
    ///
    /// # Errors
    ///
    /// Returns [`HandlerError`] when the payload cannot be decoded.
    pub fn get(&self, context: &Context, instance: &str) -> HandlerResult<Option<T>>
    where
        T: Value,
    {
        context.get_attribute_map(self, instance)
    }

    /// Reads one instance and treats absence as an error.
    ///
    /// # Errors
    ///
    /// Returns [`HandlerError`] when the instance is missing or cannot be decoded.
    pub fn get_required(&self, context: &Context, instance: &str) -> HandlerResult<T>
    where
        T: Value,
    {
        self.get(context, instance)?.ok_or_else(|| {
            HandlerError::new(format!("attribute {}[{instance}] is missing", self.name))
        })
    }

    /// Stages one instance value for durable persistence.
    ///
    /// # Errors
    ///
    /// Returns [`HandlerError`] when the value cannot be encoded or indexed.
    pub fn set(&self, context: &mut Context, instance: &str, value: T) -> HandlerResult<()>
    where
        T: Value,
    {
        context.set_attribute_map(self, instance, value)
    }

    /// Stages deletion of one instance; missing instances are valid.
    pub fn delete(&self, context: &mut Context, instance: &str) -> HandlerResult<()> {
        context.delete_attribute_map(self, instance)
    }

    /// Creates a lock request scoped to one map instance.
    pub fn lock(&self, instance: impl Into<String>) -> AttributeLock {
        AttributeLock {
            attribute: self.name.clone(),
            instance: Some(instance.into()),
        }
    }

    pub(crate) fn name(&self) -> &str {
        &self.name
    }

    pub(crate) fn index(&self) -> Option<&AttributeIndex> {
        self.index.as_ref()
    }
}

impl<T> Clone for AttributeMap<T> {
    fn clone(&self) -> Self {
        Self {
            name: self.name.clone(),
            index: self.index.clone(),
            marker: PhantomData,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
/// Describes how Dex projects an Attribute value into its Flow-search index.
///
/// Without [`Self::with_key`], a single Attribute uses its name as the physical key. Attribute maps
/// derive per-instance physical keys unless an explicit key is supplied.
pub struct AttributeIndex {
    kind: AttributeIndexKind,
    key: Option<String>,
}

impl AttributeIndex {
    fn new(kind: AttributeIndexKind) -> Self {
        Self { kind, key: None }
    }

    /// Indexes one exact UTF-8 string for equality and aggregation queries.
    pub fn keyword() -> Self {
        Self::new(AttributeIndexKind::Keyword)
    }

    /// Indexes analyzed UTF-8 text for full-text queries.
    pub fn full_text() -> Self {
        Self::new(AttributeIndexKind::FullText)
    }

    /// Indexes an array of exact UTF-8 strings.
    pub fn keyword_array() -> Self {
        Self::new(AttributeIndexKind::KeywordArray)
    }

    /// Indexes a signed 64-bit integer.
    pub fn int() -> Self {
        Self::new(AttributeIndexKind::Int)
    }

    /// Indexes a finite 64-bit floating-point number.
    pub fn double() -> Self {
        Self::new(AttributeIndexKind::Double)
    }

    /// Indexes a Boolean value.
    pub fn bool() -> Self {
        Self::new(AttributeIndexKind::Bool)
    }

    /// Indexes an RFC 3339 date-time value.
    pub fn date_time() -> Self {
        Self::new(AttributeIndexKind::DateTime)
    }

    /// Uses an explicit physical search-index key instead of the default Attribute-derived key.
    pub fn with_key(mut self, key: impl Into<String>) -> Self {
        self.key = Some(key.into());
        self
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(crate) enum AttributeIndexKind {
    Keyword,
    FullText,
    KeywordArray,
    Int,
    Double,
    Bool,
    DateTime,
}

#[derive(Clone, Debug, Eq, PartialEq)]
/// Identifies an Attribute or Attribute-map instance lock.
///
/// Create locks with [`Attribute::lock`] or [`AttributeMap::lock`], then attach them to
/// [`crate::StepOptions`] or [`crate::RpcDefinition`](crate::Rpc).
pub struct AttributeLock {
    attribute: String,
    instance: Option<String>,
}

impl AttributeIndex {
    pub(crate) fn proto_config(&self, dynamic: bool) -> IndexConfig {
        IndexConfig {
            enable: true,
            r#type: self.kind.proto_value(),
            index_key: self
                .key
                .clone()
                .or_else(|| dynamic.then(String::new))
                .unwrap_or_default(),
        }
    }

    pub(crate) fn key(&self) -> Option<&str> {
        self.key.as_deref()
    }

    pub(crate) fn proto_value(&self) -> i32 {
        self.kind.proto_value()
    }
}

impl AttributeIndexKind {
    pub(crate) fn proto_value(self) -> i32 {
        use dex_protocol::dex::IndexType;

        match self {
            Self::Keyword => IndexType::Keyword.into(),
            Self::FullText => IndexType::Text.into(),
            Self::KeywordArray => IndexType::KeywordArray.into(),
            Self::Int => IndexType::Int.into(),
            Self::Double => IndexType::Double.into(),
            Self::Bool => IndexType::Bool.into(),
            Self::DateTime => IndexType::Datetime.into(),
        }
    }
}

impl AttributeLock {
    pub(crate) fn physical_name(&self) -> String {
        match self.instance.as_deref() {
            Some(instance) => crate::registry::physical_name(&self.attribute, instance),
            None => self.attribute.clone(),
        }
    }
}
